File: src/watermelon_finder.cpp
#include "watermelon_finder.hpp"

using std::vector;
using std::make_tuple;

cv_bridge::CvImagePtr img;

bool found_watermelon() {
  return false;
}

void callback_camera(const sensor_msgs::Image::ConstPtr& img_msg) {
#if IMG_DEBUG
  ROS_INFO("Getting image");
#endif
  img = cv_bridge::toCvCopy(img_msg, "bgr8");
#if IMG_DEBUG
  cv::namedWindow("Robot View", cv::WINDOW_AUTOSIZE);
  cv::imshow("Robot View", img->image);
  cv::waitKey(0);
  cv::destroyWindow("Robot View");
  ROS_INFO("Showed the image");
#endif
}

std::ostream &operator<<(std::ostream &os, const Goal &g) {
  os << "(" << std::get<0>(g) << ", " << std::get<1>(g) << ") (" <<
    std::get<2>(g) << ", " << std::get<3>(g) << ")";
  return os;
}

int main(int argc, char** argv){
  ros::init(argc, argv, "watermelon_finder");

  //use to process images
  ros::NodeHandle nh;
  ros::Subscriber cam_sub = nh.subscribe("/camera/rgb/image_raw", 1, callback_camera);

  //tell the action client that we want to spin a thread by default
  MoveBaseClient ac("move_base", true);

  //wait for the action server to come up
  while(!ac.waitForServer(ros::Duration(5.0))){
    ROS_INFO("Waiting for the move_base action server to come up");
  }

  move_base_msgs::MoveBaseGoal goal;
  vector<Goal> goals;
  goals.push_back(make_tuple(-6.38, -1.5, 1.0, 0.0));
  goals.push_back(make_tuple(4.05, -4.43, 1.0, 0.0));
  goals.push_back(make_tuple(5.42, 2.72, -0.7, 0.7));

  for(auto it = goals.begin(); it != goals.end(); ++it) {
    goal.target_pose.header.frame_id = "map";
    goal.target_pose.header.stamp = ros::Time::now();
    
    std::tie(goal.target_pose.pose.position.x,
	     goal.target_pose.pose.position.y,
	     goal.target_pose.pose.orientation.z,
	     goal.target_pose.pose.orientation.w) = *it;

    ROS_INFO("Sending goal");
    ac.sendGoal(goal);

    ac.waitForResult();

    if(ac.getState() == actionlib::SimpleClientGoalState::SUCCEEDED) {
      ROS_INFO("The base successfully moved to (%.2f, %.2f)", std::get<0>(*it), std::get<1>(*it));
      ros::spinOnce();
      ROS_INFO("%s", found_watermelon()?"True":"False");
    }
    else
      ROS_INFO("The base failed to move to (%.2f, %.2f)", std::get<0>(*it), std::get<1>(*it));
    ros::Duration(1.0).sleep();
  }

  return 0;
}

File: src/watermelon_finder.py
#! /usr/bin/env python

import rospy
import actionlib
from actionlib_msgs.msg import *
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
from sensor_msgs.msg import Image
import tensorflow as tf
import numpy as np
import sys
import cv2

CATEGORIES = ["grapes", "pineapple","watermelon"]

def prepare(img):
    IMG_SIZE = 100 
    new_array = cv2.resize(img, (IMG_SIZE, IMG_SIZE))
    return new_array.reshape(-1, IMG_SIZE, IMG_SIZE, 3)

def callback_camera(image_data):
	rospy.loginfo("Getting image")


	img = np.frombuffer(image_data.data, dtype=np.uint8).reshape(image_data.height, image_data.width, -1)
	img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
	#cv2.namedWindow("Robot View", cv2.WINDOW_AUTOSIZE)
	#cv2.imshow("Robot View", img)
	#cv2.waitKey(0)
	#cv2.destroyWindow("Robot View")
	#rospy.loginfo("Showed the image")

	model = tf.keras.models.load_model("bestcolored_cnn.model")

	prediction = model.predict([prepare(img)])
	predicted_fruit = CATEGORIES[int(np.argmax(prediction, axis=-1))]
	if (predicted_fruit == "watermelon"):
		return True
	else:
		return False

def camera_subscriber(img):
	return

if __name__ == '__main__':
	try:

		rospy.init_node('watermelon_finder')
	
		rospy.Subscriber("/camera/rgb/image_raw", Image, camera_subscriber)
		client = actionlib.SimpleActionClient('move_base',MoveBaseAction)
		client.wait_for_server()

		goal = MoveBaseGoal()
		
		# watermelon -6.38,-1.5, 1.0, 0.0....-7.16,-2.18, 1.0, -0.08
		#pineapple 
		goal_coords = [[4.05, -4.43, 1.0, 0.0],[-7.16,-2.18, 1.0, -0.08],[5.42, 2.72, -0.7, 0.7]]
		msg = None
		for x,y,z,w in goal_coords:
			goal.target_pose.header.frame_id = "map"
			goal.target_pose.header.stamp = rospy.Time.now()
			goal.target_pose.pose.position.x = x
			goal.target_pose.pose.position.y = y
			goal.target_pose.pose.orientation.z = z
			goal.target_pose.pose.orientation.w = w

			print("Sending goal");
			client.send_goal(goal)
			client.wait_for_result()
			
			if (client.get_state() == GoalStatus.SUCCEEDED):
				rospy.loginfo("The base successfully moved to (%.2f, %.2f)", x, y)
				msg = rospy.wait_for_message("/camera/rgb/image_raw", Image)
				value = callback_camera(msg)
				if (value == True):
					print("Found a watermelon! :)")
				else:
					print("No watermelon here :(")
			else:
				rospy.loginfo("The base failed to move to (%.2f, %.2f)", x, y)
			rospy.sleep(2)
	except rospy.ROSInterruptException:
		rospy.loginfo("Terminated")
	
